src/clr-angular/data/datagrid/datagrid.ts

```diff
@@ -241,59 +236,85 @@ export class ClrDatagrid<T = any> implements AfterContentInit, AfterViewInit, On
     this._subscriptions.push(
-      this.stateProvider.change.subscribe(state => this.refresh.emit(state)),
-      this.selection.change.subscribe(s => {
-        if (this.selection.selectionType === SelectionType.Single) {
-          this.singleSelectedChanged.emit(<T>s);
-        } else if (this.selection.selectionType === SelectionType.Multi) {
-          this.selectedChanged.emit(<T[]>s);
-        }
-      }),
-      this.page.change.subscribe(() => {
-        this.datagridTable.nativeElement.focus();
-      }),
-      // A subscription that listens for displayMode changes on the datagrid
-      this.displayMode.view.subscribe(viewChange => {
-        // Remove any projected columns from the projectedDisplayColumns container
-        for (let i = this._projectedDisplayColumns.length; i > 0; i--) {
-          this._projectedDisplayColumns.detach();
-        }
-        // Remove any projected columns from the projectedCalculationColumns container
-        for (let i = this._projectedCalculationColumns.length; i > 0; i--) {
-          this._projectedCalculationColumns.detach();
-        }
-        // Remove any projected rows from the calculationRows container
-        for (let i = this._calculationRows.length; i > 0; i--) {
-          this._calculationRows.detach();
-        }
-        // Remove any projected rows from the displayedRows container
-        for (let i = this._displayedRows.length; i > 0; i--) {
-          this._displayedRows.detach();
-        }
-        if (viewChange === DatagridDisplayMode.DISPLAY) {
-          // Set state, style for the datagrid to DISPLAY and insert row & columns into containers
-          this.renderer.removeClass(this.el.nativeElement, 'datagrid-calculate-mode');
-          this.columns.forEach(column => {
-            this._projectedDisplayColumns.insert(column._view);
-          });
-          this.rows.forEach(row => {
-            this._displayedRows.insert(row._view);
-          });
-        } else {
-          // Set state, style for the datagrid to CALCULATE and insert row & columns into containers
-          this.renderer.addClass(this.el.nativeElement, 'datagrid-calculate-mode');
-          this.columns.forEach(column => {
-            this._projectedCalculationColumns.insert(column._view);
-          });
-          this.rows.forEach(row => {
-            this._calculationRows.insert(row._view);
-          });
-        }
-      })
+      this.emitOnStateProviderChange(),
+      this.emitOnSelectionChange(),
+      this.focusDgTableOnPageChange(),
+      this.resetViewsOnDisplayModeChange(),
+      this.reorderOnRequest()
     );
   }
 
+  private focusDgTableOnPageChange() {
+    return this.page.change.subscribe(() => {
+      this.datagridTable.nativeElement.focus();
+    });
+  }
+
+  private emitOnStateProviderChange(): Subscription {
+    return this.stateProvider.change.subscribe(state => this.refresh.emit(state));
```

Comment:
nit: when the function argument is matching the subscribe we could skip the function call
```ts
return this.stateProvider.change.subscribe(this.refresh.emit);
```